test/test_io.py

```diff
@@ -64,234 +64,290 @@ def temp_video(num_frames, height, width, fps, lossless=False, video_codec=None,
+                    reason="video_reader backend not available")
+@pytest.mark.skipif(av is None, reason="PyAV unavailable")
+def test_write_read_video():
+    with temp_video(10, 300, 300, 5, lossless=True) as (f_name, data):
+        lv, _, info = io.read_video(f_name)
+        assert_equal(data, lv)
+        assert info["video_fps"] == 5
+
+
+@pytest.mark.skipif(get_video_backend() != "pyav" and not io._HAS_VIDEO_OPT,
+                    reason="video_reader backend not available")
+@pytest.mark.skipif(av is None, reason="PyAV unavailable")
+@pytest.mark.skipif(not io._HAS_VIDEO_OPT, reason="video_reader backend is not chosen")
+def test_probe_video_from_file():
+    with temp_video(10, 300, 300, 5) as (f_name, data):
+        video_info = io._probe_video_from_file(f_name)
+        assert pytest.approx(video_info.video_duration, 0.1) == 2
+        assert pytest.approx(video_info.video_fps, 0.1) == 5
+
+
+@pytest.mark.skipif(get_video_backend() != "pyav" and not io._HAS_VIDEO_OPT,
+                    reason="video_reader backend not available")
+@pytest.mark.skipif(av is None, reason="PyAV unavailable")
+@pytest.mark.skipif(not io._HAS_VIDEO_OPT, reason="video_reader backend is not chosen")
+def test_probe_video_from_memory():
+    with temp_video(10, 300, 300, 5) as (f_name, data):
+        with open(f_name, "rb") as fp:
+            filebuffer = fp.read()
+        video_info = io._probe_video_from_memory(filebuffer)
+        assert pytest.approx(video_info.video_duration, 0.1) == 2
+        assert pytest.approx(video_info.video_fps, 0.1) == 5
+
+
+@pytest.mark.skipif(get_video_backend() != "pyav" and not io._HAS_VIDEO_OPT,
+                    reason="video_reader backend not available")
+@pytest.mark.skipif(av is None, reason="PyAV unavailable")
+def test_read_timestamps():
+    with temp_video(10, 300, 300, 5) as (f_name, data):
+        pts, _ = io.read_video_timestamps(f_name)
+        # note: not all formats/codecs provide accurate information for computing the
+        # timestamps. For the format that we use here, this information is available,
+        # so we use it as a baseline
+        container = av.open(f_name)
+        stream = container.streams[0]
+        pts_step = int(round(float(1 / (stream.average_rate * stream.time_base))))
+        num_frames = int(round(float(stream.average_rate * stream.time_base * stream.duration)))
+        expected_pts = [i * pts_step for i in range(num_frames)]
+
+        assert pts == expected_pts
+        container.close()
+
+
+@pytest.mark.skipif(get_video_backend() != "pyav" and not io._HAS_VIDEO_OPT,
+                    reason="video_reader backend not available")
+@pytest.mark.skipif(av is None, reason="PyAV unavailable")
+def test_read_partial_video():
+    with temp_video(10, 300, 300, 5, lossless=True) as (f_name, data):
+        pts, _ = io.read_video_timestamps(f_name)
+        for start in range(5):
+            for offset in range(1, 4):
```

Comment:
How long do these 2 lines take:

```py
    with temp_video(10, 300, 300, 5, lossless=True) as (f_name, data):
        pts, _ = io.read_video_timestamps(f_name)
```
?

If they're very short, like less than a few hundreds of ms, then we can afford to repeat those across calls and turn the 2 nested for loops into a parametrization, like so:

```py
@pytest.mark.parametrize('start', range(5))
@pytest.mark.parametrize('offset', range(1, 4))
def test_read_partial_video(start, offset):
```

I'd suggest to write the parametrization, run this test, and if the difference between the parametrized and the non-parametrized versions is small, then let's go for it.